src/callback_tracking.jl

```diff
@@ -101,6 +101,12 @@ function (f::TrackedAffect)(integrator, event_idx = nothing)
     end
 end
 
+# extend the DiffEqCallbacks.functioncalling_affect_initialize to fit TrackedAffect
+import DiffEqCallbacks: functioncalling_affect_initialize
+function functioncalling_affect_initialize(affect::TrackedAffect, u, t, integrator)
+    functioncalling_affect_initialize(affect.affect!, u, t, integrator)
+end
```

Comment:
Why would this be specialized to just a single initialize? Why not directly handle initialize!?

Reply:
I see, using a user-defined custom initialization function will suffer from the same issue. I will change both PRs, so that every `initialize!` (and also `finalize` if I am already on it?) call is filtered to `FunctionCallingAffect` instead of `TrackedAffect`

Reply:
What would be the right location to accomplish that?
One way would be to add a dispatch [in the callbacks' constructors](https://github.com/SciML/SciMLBase.jl/blob/587cfcc9cf4277b35d3a0a2b08a21d5c9467bf8f/src/callbacks.jl#L94) and substitude the given function `initialize` (and `finalize` analogous) with `(c, u, t, integrator) -> initialize(c.affect!, u, t, integrator)`  if the constructor is called with `F2` = `typeof(affect!)`  = `TrackedAffect`?

These new constructor overloads could further be deployed in SciMLSensitivity next to the definition of the `TrackedAffect`.

Reply:
Hey @ChrisRackauckas,

I took a few minutes to think about this, but adding an additional dispatch for `initialize!` ([Code](https://github.com/SciML/DiffEqBase.jl/blob/e8c0a33c5a6be7d7bbbca190e7e0a14d5a5f7f1d/src/callbacks.jl#L16)) may not be the right way. The function interface of the callback initialization function is `(c::DECallback, u, t, integrator)`, so another dispatch that invokes the function on `c.affect!` would violate this interface. Initialization routines will be able to use `c.affect!` as to expect, but other attributes of `FunctionCallingAffect` are not part of the `TrackedAffect`, like for example `funciter`.

Curren timplementation, see [Code](https://github.com/SciML/DiffEqBase.jl/blob/e8c0a33c5a6be7d7bbbca190e7e0a14d5a5f7f1d/src/callbacks.jl#L16):
```
function initialize!(u, t, integrator::DEIntegrator, any_modified::Bool,
                     c::DECallback)
    c.initialize(c, u, t, integrator)
    any_modified || integrator.u_modified
end
```

As far as I understand, this would be the method to add (but it's bad!):
(see [Code](https://github.com/SciML/SciMLBase.jl/blob/f45dd60b50a9c207e9631356e887522d210a0d73/src/callbacks.jl#L293) for the DiscreteCallback implementation)
```
# new dispatch for discrete callbacks with `typeof(affect!)==TrackedAffect`
function initialize!(u, t, integrator::DEIntegrator, any_modified::Bool,
                     c::DiscreteCallback{Any, TrackedAffect, Any, Any}) 
    c.initialize(c.affect!, u, t, integrator) # we need to dive on layer into the TrackedAffect
    any_modified || integrator.u_modified
end
```

The problem here is, that the initialization routine is not performed on the callback itself, but on the `TrackedAffect`. So this will not work.

Because the core problem is that a TrackedAffect doesn't behave like a FunctionCallingAffect (because of the different fields), my suggestion would be to mimic the existence of the fields of the  FunctionCallingAffect inside of the TrackedAffect struct - this could easily be implemented overloading the `hasproperty`, `setproperty!` and `getproperty` functions. Another advantage of this is, that all changes would be inside of DiffEqCallbacks and other libraries are not affected.

What do you think?
I can rapidely deploy code if this is the way to go.

Best regards, 
Tobi

Reply:
I think it may need to be a separate TrackedAffect. Thoughts @frankschae?

Reply:
Wouldn't it work if we extend the signature for `initialize!` to depend on both the callback and the `TrackedAffect`? (Not sure we really want to do it that way.... but it seems much better to handle different custom initializations as well.)

Reply:
yeah, this seems like a more general way to handle it to me.

Reply:
I will make a quick suggestion, with `initialize!` depending on the callback and the (correct) `affect!`.